service/history/api/recordchildworkflowcompleted/api.go

```diff
@@ -93,11 +93,29 @@ func Invoke(
 
 			// Check mutable state to make sure child execution is in pending child executions
 			ci, isRunning := mutableState.GetChildExecutionInfo(parentInitiatedID)
-			if !isRunning || ci.StartedEventId == common.EmptyEventID {
-				// note we already checked if startedEventID is empty (in consistency predicate)
-				// and reloaded mutable state
+			if !isRunning {
 				return nil, consts.ErrChildExecutionNotFound
 			}
+			if ci.StartedEventId == common.EmptyEventID {
+				// note we already checked if startedEventID is empty (in consistency predicate)
+				// and reloaded mutable state, so if startedEventID is still missing, we need to
+				// record a started event before recording completion event.
+				initiatedEvent, err := mutableState.GetChildExecutionInitiatedEvent(ctx, parentInitiatedID)
+				if err != nil {
+					return nil, consts.ErrChildExecutionNotFound
+				}
+				initiatedAttr := initiatedEvent.GetStartChildWorkflowExecutionInitiatedEventAttributes()
+				_, err = mutableState.AddChildWorkflowExecutionStartedEvent(
+					request.GetChildExecution(),
+					initiatedAttr.WorkflowType,
+					initiatedEvent.EventId,
+					initiatedAttr.Header,
+					ci.Clock,
```

Comment:
I think ci.Clock is not initialized here (maybe use nil explicitly?). But I guess that doesn't matter. Once child response is recorded, the child info will be deleted, so we can use any arbitrary here.

Can you add some comments here?